### PR TITLE
Feature/ssr-confirm

### PR DESCRIPTION
**Pull request by UmttikhinaDasha**
- move confirmation pages (EmpInvite, LinkInvite, Activation, EmailActivation, EmailChange, PasswordChange, ProfileRemove, PhoneActivation, PortalOwnerChange, TfaAuth, TfaActivation, PortalRemove, PortalSuspend, PortalContinue, Auth)
- move wizard page
- change layout and style components

### CODE REVIEW

File: packages/client/src/pages/PortalSettings/categories/security/access-portal/tfa.js [hunk 1]
```diff
@@ -150,7 +150,7 @@ const TwoFactorAuth = (props) => {
 
       if (res) {
         setIsInit(false);
-        navigate(res.replace(window.location.origin, ""));
+        if (res) window.location.replace(res);
```
Comment (TimofeyBoyko) [inline packages/client/src/pages/PortalSettings/categories/security/access-portal/tfa.js:153]: Useless if(res) check.

File: packages/client/src/pages/Profile/Section/Body/sub-components/login-settings/index.js [hunk 2]
```diff
@@ -57,8 +57,9 @@ const LoginSettings = (props) => {
     useState(false);
 
   useEffect(() => {
-    if (location?.state?.openBackupCodesDialog) {
+    if (sessionStorage.getItem("openBackupCodesDialog")) {
```
Comment (TimofeyBoyko) [inline packages/client/src/pages/Profile/Section/Body/sub-components/login-settings/index.js:60]: Move 'openBackupCodesDialog' to constant and fix using

File: packages/login/src/HOCs/withLoader.tsx [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+import {
+  getCapabilities,
+  getPortalPasswordSettings,
+  getThirdPartyProviders,
+  getUserByEmail,
+} from "@/utils/actions";
+import { TError, WithLoaderProps } from "@/types";
+import { ConfirmRouteContext } from "@/components/ConfirmRoute";
+import Loading from "../app/(root)/confirm/loading";
```
Comment (TimofeyBoyko) [inline packages/login/src/HOCs/withLoader.tsx:52]: import AppLoader from shared

File: packages/login/src/HOCs/withLoader.tsx
Comment (TimofeyBoyko) [inline packages/login/src/HOCs/withLoader.tsx:1]: Maybe move this logic at server side?

Comment (TimofeyBoyko) [reply packages/login/src/HOCs/withLoader.tsx:1]: Loaded state can stay in client side, but API requests need move to server side

File: packages/login/src/app/(root)/confirm/Activation/page.tsx [hunk 4]
```diff
@@ -24,46 +24,39 @@
...
 
   return (
-    <LanguageCombobox
-      className="language-combo-box"
-      onSelectLanguage={onLanguageSelect}
-      cultures={cultures}
-      selectedCulture={currentCultureName}
-      withBorder={!isMobileView}
-    />
+    <>
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/Activation/page.tsx:47]: useless <> </>

File: packages/login/src/app/(root)/confirm/Activation/page.tsx [hunk 5]
```diff
@@ -24,46 +24,39 @@
...
-
-const LanguageComboboxWrapper = (props: TLanguageCombobox) => {
-  const { cultures, currentCultureName, currentDeviceType } = props;
+import { getSettings, getUserFromConfirm } from "@/utils/actions";
 
-  const onLanguageSelect = (culture: { key: string }) => {
-    const { key } = culture;
+import ActivateUserForm from "@/components/ActivateUserForm";
+import {
+  GreetingCleateUserContainer,
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/Activation/page.tsx:33]: Component name '...Cleate...'

File: packages/login/src/app/(root)/confirm/Activation/page.tsx [hunk 6]
```diff
@@ -24,46 +24,39 @@
...
+  GreetingContainer,
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/Activation/page.tsx:34]: Useless import

File: packages/login/src/app/(root)/confirm/PasswordChange/page.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+import { GreetingContainer } from "@/components/GreetingContainer";
+import PasswordChangeForm from "@/components/PasswordChangeForm";
+import { getSettings } from "@/utils/actions";
+
+type PasswordChangProps = {
+  searchParams: { [key: string]: string };
+};
+
+async function Page({ searchParams }: PasswordChangProps) {
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/PasswordChange/page.tsx:37]: Useless 'searchParams' prop

File: packages/login/src/app/(root)/confirm/EmailChange/page.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+// content are licensed under the terms of the Creative Commons Attribution-ShareAlike 4.0
+// International. See the License terms at http://creativecommons.org/licenses/by-sa/4.0/legalcode
+
+import EmailChangeHandler from "@/components/EmailChangeHandler";
+
+type EmailChangeProps = {
+  searchParams: { [key: string]: string };
+};
+
+async function Page({ searchParams }: EmailChangeProps) {
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/EmailChange/page.tsx:33]: Useless 'searchParams' prop

File: packages/login/src/app/(root)/confirm/EmailActivation/page.tsx [hunk 9]
```diff
@@ -24,16 +24,14 @@
...
+import EmailActivationHandler from "@/components/EmailActivationHandler";
 
-import ErrorBoundaryWrapper from "SRC_DIR/components/ErrorBoundaryWrapper";
+type EmailActivationProps = {
+  searchParams: { [key: string]: string };
+};
 
-export const Component = () => {
-  //console.log("Confirm render");
+async function Page({ searchParams }: EmailActivationProps) {
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/EmailActivation/page.tsx:33]: Useless 'searchParams' prop

File: packages/login/src/app/(root)/confirm/PhoneActivation/page.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+import { GreetingContainer } from "@/components/GreetingContainer";
+import { getSettings } from "@/utils/actions";
+import ChangePhoneForm from "@/components/ChangePhoneForm";
+
+type PhoneActivationProps = {
+  searchParams: { [key: string]: string };
+};
+
+async function Page({ searchParams }: PhoneActivationProps) {
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/PhoneActivation/page.tsx:37]: Useless 'searchParams' prop

File: packages/login/src/app/(root)/confirm/PortalContinue/page.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+import { GreetingContainer } from "@/components/GreetingContainer";
+import { getSettings } from "@/utils/actions";
+import ContinuePortalForm from "@/components/ContinuePortalForm";
+
+type PortalContinueProps = {
+  searchParams: { [key: string]: string };
+};
+
+async function Page({ searchParams }: PortalContinueProps) {
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/PortalContinue/page.tsx:37]: Useless 'searchParams' prop

File: packages/login/src/app/layout.tsx [hunk 12]
```diff
@@ -50,27 +55,24 @@ export default async function RootLayout({
...
-
   const cookieStore = cookies();
 
   const systemTheme = cookieStore.get(SYSTEM_THEME_KEY);
   const cookieLng = cookieStore.get(LANGUAGE);
 
   let redirectUrl = "";
 
-  const [settings, colorTheme] = await Promise.all([
+  const [settings, colorTheme, user] = await Promise.all([
```
Comment (TimofeyBoyko) [inline packages/login/src/app/layout.tsx:65]: Not need fetch user here

It need only for confirms page, add check

File: packages/login/src/app/wizard/page.tsx [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+async function Page() {
+  const settings = await getSettings();
+
+  const objectSettings = typeof settings === "string" ? undefined : settings;
+
+  if (!objectSettings || !objectSettings.wizardToken) {
+    redirect("/");
+  }
+
+  const passwordSettingsData = getPortalPasswordSettings(
```
Comment (TimofeyBoyko) [inline packages/login/src/app/wizard/page.tsx:47]: move all function calls to Promise.all array

File: packages/shared/components/combobox/Combobox.types.ts [hunk 14]
```diff
@@ -51,7 +51,7 @@ export type TOption =
     }
   | {
       key: string | number;
-      icon?: undefined;
+      icon?: string;
```
Comment (TimofeyBoyko) [inline packages/shared/components/combobox/Combobox.types.ts:54]: For separator option all params must be undefined

File: packages/shared/components/combobox/Combobox.types.ts [hunk 15]
```diff
@@ -60,7 +60,7 @@ export type TOption =
       disabled?: undefined;
       description?: undefined;
       quota?: undefined;
-      isSeparator: true;
+      isSeparator?: true;
```
Comment (TimofeyBoyko) [inline packages/shared/components/combobox/Combobox.types.ts:63]: This must be always true for separator option

File: packages/shared/api/settings/index.ts [hunk 16]
```diff
@@ -721,7 +723,7 @@ export function getTfaSecretKeyAndQR(confirmKey = null) {
   return request(options);
 }
 
-export function validateTfaCode(code, confirmKey = null) {
+export function validateTfaCode(code, confirmKey: string | null = null) {
```
Comment (TimofeyBoyko) [inline packages/shared/api/settings/index.ts:726]: Use Nullable<string> instead of string | null

File: packages/shared/api/portal/index.ts [hunk 17]
```diff
@@ -213,7 +213,7 @@ export function sendDeletePortalEmail() {
   });
 }
 
-export function suspendPortal(confirmKey = null) {
+export function suspendPortal(confirmKey: string | null = null) {
```
Comment (TimofeyBoyko) [inline packages/shared/api/portal/index.ts:216]: Use Nullable<string> instead of string | null

File: packages/shared/api/portal/index.ts [hunk 18]
```diff
@@ -235,7 +235,7 @@ export function continuePortal(confirmKey = null) {
   return request(options);
 }
 
-export function deletePortal(confirmKey = null) {
+export function deletePortal(confirmKey: string | null = null) {
```
Comment (TimofeyBoyko) [inline packages/shared/api/portal/index.ts:238]: Use Nullable<string> instead of string | null

File: packages/shared/api/people/index.ts [hunk 19]
```diff
@@ -129,7 +132,7 @@ export function getUserPhoto(userId) {
   });
 }
 
-export function createUser(data, confirmKey = null) {
+export function createUser(data, confirmKey: string | null = null) {
```
Comment (TimofeyBoyko) [inline packages/shared/api/people/index.ts:135]: Use Nullable<string> instead of string | null

File: packages/login/src/utils/index.ts
Comment (TimofeyBoyko) [inline packages/login/src/utils/index.ts:1]: Move this logic to shared/utils

File: packages/login/src/components/ProfileRemoveForm /index.tsx [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+import { TError } from "@/types";
+
+import { GreetingContainer } from "../GreetingContainer";
+import { ConfirmRouteContext } from "../ConfirmRoute";
+
+type ProfileRemoveFormProps = {
+  legalTerms: string;
+  greetingSettings: string;
+  currentColorScheme?: TColorScheme;
```
Comment (TimofeyBoyko) [inline packages/login/src/components/ProfileRemoveForm /index.tsx:48]: Better get it from 'useTheme' from 'styled-components'

File: packages/login/src/app/(root)/confirm/ProfileRemove/page.tsx [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+// content are licensed under the terms of the Creative Commons Attribution-ShareAlike 4.0
+// International. See the License terms at http://creativecommons.org/licenses/by-sa/4.0/legalcode
+
+import { getColorTheme, getSettings } from "@/utils/actions";
+import ProfileRemoveForm from "@/components/ProfileRemoveForm ";
+
+async function Page() {
+  const [settings, colorTheme] = await Promise.all([
+    getSettings(),
+    getColorTheme(),
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/ProfileRemove/page.tsx:33]: Remove and get color theme from 'useTheme' at client component

File: packages/login/src/app/(root)/confirm/TfaActivation/page.tsx [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+  getTfaSecretKeyAndQR,
+  getUserFromConfirm,
+} from "@/utils/actions";
+
+type TfaActivationProps = {
+  searchParams: { [key: string]: string };
+};
+
+async function Page({ searchParams }: TfaActivationProps) {
+  const colorTheme = await getColorTheme();
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/confirm/TfaActivation/page.tsx:42]: Remove and get color theme from 'useTheme' at client component 

Or move it to Promise.all

File: packages/login/src/app/(root)/page.tsx [hunk 23]
```diff
@@ -33,12 +33,16 @@ import {
...
 } from "@/utils/actions";
-import Login from "@/components/Login";
+import Login, { LoginContent } from "@/components/Login";
 import LoginForm from "@/components/LoginForm";
 import ThirdParty from "@/components/ThirdParty";
 import RecoverAccess from "@/components/RecoverAccess";
 import Register from "@/components/Register";
+import { GreetingLoginContainer } from "@/components/GreetingContainer";
 import { FormWrapper } from "@docspace/shared/components/form-wrapper";
+import LanguageComboboxWrapper from "@/components/LanguageCombobox";
```
Comment (TimofeyBoyko) [inline packages/login/src/app/(root)/page.tsx:43]: Missed dynamic import and disabled ssr (see old version)

File: packages/login/src/app/wizard/layout.tsx [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+  const cookieStore = cookies();
+
+  const systemTheme = cookieStore.get(SYSTEM_THEME_KEY)?.value as ThemeKeys;
+
+  const bgPattern = getBgPattern(colorTheme?.selected);
+
+  const objectSettings = typeof settings === "string" ? undefined : settings;
+
+  const isRegisterContainerVisible = objectSettings?.enabledJoin;
```
Comment (TimofeyBoyko) [inline packages/login/src/app/wizard/layout.tsx:63]: Register container at wizard?

File: packages/shared/api/portal/index.ts [hunk 25]
```diff
@@ -33,6 +33,7 @@ import {
   TPortalTariff,
   TRestoreProgress,
 } from "./types";
+import { Nullable } from "types";
```
Comment (TimofeyBoyko) [inline packages/shared/api/portal/index.ts:36]: Change import path to '../../types'

File: packages/shared/api/people/index.ts [hunk 26]
```diff
@@ -30,6 +30,7 @@ import { AccountsSearchArea } from "@docspace/shared/enums";
 // import axios from "axios";
 import { Encoder } from "@docspace/shared/utils/encoder";
 import { checkFilterInstance } from "@docspace/shared/utils/common";
+import { Nullable } from "types";
```
Comment (TimofeyBoyko) [inline packages/shared/api/people/index.ts:33]: change import path to '../../types'

File: packages/login/src/types/index.ts [hunk 27]
```diff
@@ -24,27 +25,154 @@
...
+  account: string;
+  manualEntryKey: string;
+  qrCodeSetupImageUrl: string;
+};
+
+export interface ConfirmRouteProps {
+  doAuthenticated?: AuthenticatedAction;
+  defaultPage?: string;
+  socketUrl?: string;
+  children: ReactNode | string;
```
Comment (TimofeyBoyko) [inline packages/login/src/types/index.ts:169]: ReactNode include string

File: packages/login/src/providers/index.tsx [hunk 28]
```diff
@@ -41,29 +41,43 @@ import useTheme from "@/hooks/useTheme";
...
   redirectURL: string;
+  user?: TUser;
 }) => {
   const firebaseHelper = new FirebaseHelper(
     value.settings?.firebase ?? ({} as TFirebaseSettings),
   );
+  const confirmType = useSearchParams().get("type");
+
+  let shouldRedirect = true;
+  if (redirectURL == `unavailable` && confirmType == "PortalContinue") {
```
Comment (TimofeyBoyko) [inline packages/login/src/providers/index.tsx:63]: Change to '==='

File: packages/login/src/components/WizardForm/WizardForm.styled.ts [hunk 29]
```diff
@@ -159,19 +110,22 @@ export const StyledInfo = styled.div`
...
 
-  .text {
-    color: ${(props) => props.theme.client.wizard.textColor};
-  }
-
   .machine-name {
     padding-bottom: 4px;
     padding-top: 4px;
-    padding-inline-start: 8px;
+    padding-left: 8px;
```
Comment (TimofeyBoyko) [inline packages/login/src/components/WizardForm/WizardForm.styled.ts:116]: Return padding-inline-start

File: packages/login/src/components/TfaActivationForm/index.tsx [hunk 30]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  };
+
+  const onChangeInput = (event: ChangeEvent<HTMLInputElement>) => {
+    setCode(event.target.value);
+    setError("");
+  };
+
+  const onKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.code === "Enter" || event.code === "NumpadEnter") onSubmit();
```
Comment (TimofeyBoyko) [inline packages/login/src/components/TfaActivationForm/index.tsx:128]: Use ButtonKeys enum

File: packages/login/src/components/ProfileRemoveForm /index.tsx [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+}: ProfileRemoveFormProps) => {
+  const { linkData } = useContext(ConfirmRouteContext);
+  const { t } = useTranslation(["Confirm", "Common"]);
+  const [isProfileDeleted, setIsProfileDeleted] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
+
+  const onDeleteProfile = () => {
+    setIsLoading(true);
+
+    deleteSelf(linkData.confirmHeader)
```
Comment (TimofeyBoyko) [inline packages/login/src/components/ProfileRemoveForm /index.tsx:64]: Use await\async

File: packages/login/src/components/PasswordChangeForm/index.tsx [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+        toastr.error(t("Common:SomethingWentWrong"));
+      } else {
+        toastr.error(t(`${errorMessage}`));
+      }
+      setIsLoading(false);
+    }
+  };
+
+  const onKeyPress = (e: KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === "Enter") {
```
Comment (TimofeyBoyko) [inline packages/login/src/components/PasswordChangeForm/index.tsx:139]: Use ButtonKeys enum

File: packages/login/src/components/GreetingContainer/index.tsx [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+//
+// Pursuant to Section 7(b) of the License you must retain the original Product logo when
+// distributing the program. Pursuant to Section 7(e) we decline to grant you any rights under
+// trademark law for use of our trademarks.
+//
+// All the Product's GUI elements, including illustrations and icon sets, as well as technical writing
+// content are licensed under the terms of the Creative Commons Attribution-ShareAlike 4.0
+// International. See the License terms at http://creativecommons.org/licenses/by-sa/4.0/legalcode
+
+export { GreetingCreateUserContainer } from "./GreetingCreateUserContainer/GreetingCreateUserContainer";
```
Comment (TimofeyBoyko) [inline packages/login/src/components/GreetingContainer/index.tsx:27]: Change path to "./GreetingCreateUserContainer" (rename GreetingCreateUserContainer.tsx file to index.tsx)

File: packages/login/src/components/EmailChangeHandler/index.tsx [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+import { ConfirmRouteContext } from "../ConfirmRoute";
+
+const EmailChangeHandler = () => {
+  const [error, setError] = useState<string>();
+
+  const { linkData } = useContext(ConfirmRouteContext);
+  const { email = "", uid = "", key = "" } = linkData;
+
+  useEffect(() => {
+    changeEmail(uid, email, key)
```
Comment (TimofeyBoyko) [inline packages/login/src/components/EmailChangeHandler/index.tsx:45]: Use async\await

File: packages/login/src/components/EmailActivationHandler/index.tsx [hunk 35]
```diff
@@ -0,0 +1,79 @@
...
+const EmailActivationHandler = () => {
+  const [error, setError] = useState<string>();
+
+  const { linkData } = useContext(ConfirmRouteContext);
+  const router = useRouter();
+
+  const { email, uid = "", key = "" } = linkData;
+
+  useEffect(() => {
+    updateActivationStatus(EmployeeActivationStatus.Activated, uid, key)
```
Comment (TimofeyBoyko) [inline packages/login/src/components/EmailActivationHandler/index.tsx:49]: use async\await

File: packages/login/src/components/CreateUserForm/index.tsx [hunk 36]
```diff
@@ -0,0 +1,529 @@
...
+    async (profile: string) => {
+      const signupAccount: { [key: string]: string } = {
+        EmployeeType: linkData.emplType ?? "",
+        Email: linkData.email ?? "",
+        Key: linkData.key ?? "",
+        SerializedProfile: profile,
+        culture: currentCultureName,
+      };
+
+      return signupOAuth(signupAccount)
```
Comment (TimofeyBoyko) [inline packages/login/src/components/CreateUserForm/index.tsx:154]: Use async/await

File: packages/login/src/components/CreateUserForm/index.tsx [hunk 37]
```diff
@@ -0,0 +1,529 @@
...
+      try {
+        const tokenGetterWin = isDesktopView
+          ? (window.location.href = url)
+          : window.open(
+              url,
+              "login",
+              "width=800,height=500,status=no,toolbar=no,menubar=no,resizable=yes,scrollbars=no",
+            );
+
+        getOAuthToken(tokenGetterWin).then((code) => {
```
Comment (TimofeyBoyko) [inline packages/login/src/components/CreateUserForm/index.tsx:411]: use async/await

File: packages/login/src/components/CreateUserForm/index.tsx [hunk 38]
```diff
@@ -0,0 +1,529 @@
...
+      confirmUser.type = type;
+    }
+
+    if (!!linkData.key) {
+      confirmUser.key = linkData.key;
+    }
+
+    const headerKey = linkData?.confirmHeader ?? "";
+
+    createConfirmUser(confirmUser, headerKey).catch((error) => {
```
Comment (TimofeyBoyko) [inline packages/login/src/components/CreateUserForm/index.tsx:302]: use async/await

File: packages/login/src/components/AuthHandler/index.tsx [hunk 39]
```diff
@@ -24,31 +24,37 @@
...
+import { ConfirmRouteContext } from "../ConfirmRoute";
+
+const AuthHandler = () => {
+  let searchParams = useSearchParams();
+
+  const { linkData } = useContext(ConfirmRouteContext);
+  const { email = "", key = "" } = linkData;
+
   useEffect(() => {
     loginWithConfirmKey({
```
Comment (TimofeyBoyko) [inline packages/login/src/components/AuthHandler/index.tsx:51]: use async/await

File: packages/login/src/components/ActivateUserForm/index.tsx [hunk 40]
```diff
@@ -0,0 +1,300 @@
...
+    const { userName, passwordHash } = loginData;
+
+    const res1 = await changePassword(userId, passwordHash, key);
+    const res2 = await updateActivationStatus(activationStatus, userId, key);
+    const res3 = await login(userName, passwordHash);
+    const res4 = await updateUser(changedData);
+  };
+
+  const onKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === "Enter") {
```
Comment (TimofeyBoyko) [inline packages/login/src/components/ActivateUserForm/index.tsx:187]: use enum

File: packages/login/src/components/ActivateUserForm/index.tsx [hunk 41]
```diff
@@ -0,0 +1,300 @@
...
+  }: TActivateConfirmUser) => {
+    const changedData = {
+      id: userId,
+      FirstName: personalData.firstname,
+      LastName: personalData.lastname,
+    };
+
+    const { userName, passwordHash } = loginData;
+
+    const res1 = await changePassword(userId, passwordHash, key);
```
Comment (TimofeyBoyko) [inline packages/login/src/components/ActivateUserForm/index.tsx:180]: try promise.all

Comment (UmttikhinaDasha) [reply packages/login/src/components/ActivateUserForm/index.tsx:180]: promise.all does not work because these requests must be executed one after another

**Review by TimofeyBoyko — APPROVED** [review]

**Review by AlexeySafronov — APPROVED** [review]

**Review by TimofeyBoyko — APPROVED** [review]

**Pull request merged**